### PR TITLE
@dzucconi: Can properly query "my active" bids (DO NOT MERGE YET)

### PR DESCRIPTION
**Pull request by craigspaeth**
I still need to write tests, but wanted to put this up for review. I'll comment inline...

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -34,7 +34,7 @@
     "express-graphql": "^0.4.4",
     "graphiql": "^0.3.1",
     "graphql": "^0.4.13",
-    "lodash": "^3.10.1",
+    "lodash": "^4.3.0",
```
Comment (craigspaeth) [inline package.json]: Updated to lodash 4 to get `uniqBy`—there's a [lot of breaking changes](https://github.com/lodash/lodash/wiki/Changelog#v400), but I tried to search the project to cover them. Tests pass, but happy to punt and make this a separate PR (`uniq` in 3.0 would work fine).

File: schema/bidder_position.js [hunk 2]
```diff
@@ -45,6 +45,15 @@ const BidderPositionType = new GraphQLObjectType({
       type: SaleArtwork.type,
       resolve: position => gravity(`sale_artwork/${position.sale_artwork_id}`),
     },
+    is_winning: {
```
Comment (craigspaeth) [inline schema/bidder_position.js]: Not sure if this it the best way to go about this. This field would mimic [the Force PR logic](https://github.com/artsy/force/pull/4284/files#diff-e13e2d4daed3c7935d99823410845942R4).

It seems silly if this fetches the sale artwork twice when used in combo with `sale_artwork`, e.g.

````
query {
  me {
    bidder_positions(current: true) {
      id
      is_winning
      sale_artwork {
        id
      }
    }
  }
}
````

Although I wonder if that's a bid deal depending on how the caching works. I also wonder if there's some kind of currying/lazy-loading thing we could put in `gravity` that can recognize the same query being made twice at the same time and make only one fetch.

Curious to hear your thoughts on this implementation.

File: schema/bidder_positions.js [hunk 3]
```diff
@@ -1,15 +0,0 @@
-import gravity from '../lib/loaders/gravity';
-import BidderPosition from './bidder_position';
-import {
-  GraphQLList,
-} from 'graphql';
-
-const BidderPositions = {
```
Comment (craigspaeth) [inline schema/bidder_positions.js]: Moved this into the me schema, wasn't doing much.

File: schema/image/resized.js [hunk 4]
```diff
@@ -17,7 +17,7 @@ export const resizedImageUrl = (image, options) => {
   const desired = _.pick(opts, 'width', 'height');
   const factor = _.min(_.map(desired, (value, attr) => {
     return value / image[`original_${attr}`];
-  }));
+  })) || Infinity;
```
Comment (craigspaeth) [inline schema/image/resized.js]: Seemed like maybe `_.min` had a different return value in lodash 4—but I can better investigate that, just dropped this in b/c it made the test happy. I dunno if it has unforeseen repercussions.

File: schema/sale_artwork.js [hunk 5]
```diff
@@ -40,17 +35,6 @@ const SaleArtworkType = new GraphQLObjectType({
       bidder_positions_count: {
         type: GraphQLInt,
       },
-      bidder_positions: {
```
Comment (craigspaeth) [inline schema/sale_artwork.js]: Removed this just b/c we aren't using it currently, but happy to keep it in.

File: schema/bidder_position.js (hunk 2 above)
Comment (dzucconi) [inline schema/bidder_position.js]: > Although I wonder if that's a bid deal depending on how the caching works. I also wonder if there's some kind of currying/lazy-loading thing we could put in gravity that can recognize the same query being made twice at the same time and make only one fetch.

That is exactly what happens actually. You should be able to debug the number of requests that's happening by looking in the console but, if you're using the gravity loader (which you are) to load a resource, then it will only be fetched a single time.

File: schema/me.js [hunk 6]
```diff
@@ -22,7 +27,24 @@ const Me = new GraphQLObjectType({
       resolve: ({ default_profile_id }) =>
         gravity(`profile/${default_profile_id}`),
     },
-    bidder_positions: BidderPositions,
+    bidder_positions: {
+      type: new GraphQLList(BidderPosition.type),
+      description: "A list of the current user's bidder positions",
+      args: {
+        current: {
```
Comment (craigspaeth) [inline schema/me.js]: `/api/v1/me/bidder_positions` returns all bids (including historical bids). For instance if Devang bids $10k on a work, then later ups his max bid to $20k—it will create 2 bidder positions, but the $10k one is the actually relavant one, until someone bids higher and the 20k bid outbids them. The way it's implemented now it can create a weird "historical" view of bids (see below). This `current` arg will allow one to filter the bidder positions to _only_ the most recent/relevant bids. This is what tripped me up yesterday and cause me to pend the my active bids.

image

File: package.json (hunk 1 above)
Comment (dzucconi) [inline package.json]: If the specs pass then that should be good. Thanks

File: schema/sale_artwork.js (hunk 5 above)
Comment (dzucconi) [inline schema/sale_artwork.js]: I suppose it depends on whether or not we want to enforce the `me`-scoped queries. I was initially planning on using this on the artwork page auction interface but I'll try coming at it from `me` first.

File: schema/me.js (hunk 6 above)
Comment (dzucconi) [inline schema/me.js]: :+1:

File: schema/image/resized.js (hunk 4 above)
Comment (dzucconi) [inline schema/image/resized.js]: Hm; yeah I'm not sure but the specs cover this object pretty well so if we pass there it should be fine.

File: schema/bidder_position.js (hunk 2 above)
Comment (craigspaeth) [inline schema/bidder_position.js]: Oh! Aweeesome!

**Comment by craigspaeth** [comment]
:sweat_smile: took a long time to figure out I needed to pass `accessToken` in to the [test here](https://github.com/artsy/metaphysics/pull/115/files#diff-5c0f7b5084a56474fd0f9b2d8062d2b4R80) otherwise it would silently fail [destructuring here](https://github.com/artsy/metaphysics/pull/115/files#diff-d699f4befea42add033250bfe1ffea9fR40). Wrote some tests and kick semaphore :green_heart: 

Do you think it's possible errors like that are silently failing in the promise world somewhere? I remember at one point  comparing Q and Bluebird and remembering that bluebird would log uncaught errors where they would silently pass in Q.

**Comment by dzucconi** [comment]
So yeah; we're using native Promises which, by default, swallow errors.

When writing specs I'll often temporarily include a `catch` so I know what I'm doing.

We could certainly just use [Bluebird](http://jamesknelson.com/are-es6-promises-swallowing-your-errors/)... in practice GraphQL does a nice job of surfacing them, and I haven't had any real complaints about this yet.

**Comment by dzucconi** [comment]
Good to merge?

**Comment by craigspaeth** [comment]
Yep yep, good to merge! Thanks for the context, I'll be sure to drop in some `catch`es next time I'm scratching my head.
